File: src/api_service/rfq.ts
import { ApiClient } from './client';

export interface RFQItem {

  id?: number;
  rfq_id?: number;
  product_category_id?: number;
  item_name: string;
  description?: string;
  quantity: number;
  unit_id: number;
  estimated_unit_price?: number;
  specifications?: string;
  sort_order: number;
  part_number?: string; // Added optional part number
  is_approved?: boolean;
}

export interface RFQAttachment {
  id: number;
  rfq_id: number;
  file_name: string;
  file_url: string;
  file_size_bytes?: number;
  uploaded_by: number;
  uploaded_at: string;
}

export interface RFQVendorInvitation {
  rfq_id: number;
  vendor_id: number;
  invited_at: string;
  notified_at?: string;
}

export interface RFQ {
  id?: number;
  rfq_number?: string;
  title: string;
  description?: string;
  status: string;
  submission_deadline: string;
  delivery_deadline?: string;
  created_by?: number;
  closed_at?: string;
  created_at?: string;
  updated_at?: string;

  items?: RFQItem[];
  attachments?: RFQAttachment[];
  invitations?: RFQVendorInvitation[];
}

export interface RFQSearchResult {
  page: number;
  limit: number;
  total: number;
  total_pages: number;
  rfqs: RFQ[];
}

export class RFQService {
  static async createRFQ(formData: FormData): Promise<{ id: number; rfq_number: string; message: string }> {
    return ApiClient.postFormData<{ id: number; rfq_number: string; message: string }>('/api/rfqs', formData);
  }

  static async searchRFQs(query: string, status?: string, page: number = 1, limit: number = 20): Promise<RFQSearchResult> {
    const params = new URLSearchParams();
    if (query) params.append('q', query);
    if (status) params.append('status', status);
    
    const offset = (page - 1) * limit;
    params.append('limit', limit.toString());
    params.append('offset', offset.toString());

    return ApiClient.get<RFQSearchResult>(`/api/rfqs/search?${params.toString()}`);
  }

  static async getRFQ(id: number): Promise<RFQ> {
    return ApiClient.get<RFQ>(`/api/rfqs/${id}`);
  }

  static async updateRFQ(id: number, formData: FormData): Promise<{ message: string }> {
    const handleSubmit = async (e: any) => {
      e.preventDefault();
      try {
        const payload = {
          email: formData.get('email'),
          company_name: formData.get('companyName'),
        };
        await ApiClient.post('/api/auth/signup', payload);
      } catch (err: any) {
        console.error(err);
      }
    };
    return ApiClient.putFormData<{ message: string }>(`/api/rfqs/${id}`, formData);
  }

  static async deleteRFQ(id: number): Promise<{ message: string }> {
    return ApiClient.request<{ message: string }>(`/api/rfqs/${id}`, {
      method: 'DELETE',
    });
  }
}

File: backend/internal/models/rfq.go
package models

import "time"

type RFQ struct {
	ID                 int64                 `json:"id"`
	RFQNumber          string                `json:"rfq_number"`
	Title              string                `json:"title"`
	Description        *string               `json:"description,omitempty"`
	Status             string                `json:"status"`
	SubmissionDeadline time.Time             `json:"submission_deadline"`
	DeliveryDeadline   *time.Time            `json:"delivery_deadline,omitempty"`
	CreatedBy          int64                 `json:"created_by"`
	ClosedAt           *time.Time            `json:"closed_at,omitempty"`
	CreatedAt          time.Time             `json:"created_at"`
	UpdatedAt          time.Time             `json:"updated_at"`
	Items              []RFQItem             `json:"items,omitempty"`
	Attachments        []RFQAttachment       `json:"attachments,omitempty"`
	Invitations        []RFQVendorInvitation `json:"invitations,omitempty"`
}

type RFQItem struct {
	ID                 int64    `json:"id"`
	RFQID              int64    `json:"rfq_id"`
	ProductCategoryID  *int     `json:"product_category_id,omitempty"`
	ItemName           string   `json:"item_name"`
	Description        *string  `json:"description,omitempty"`
	Quantity           float64  `json:"quantity"`
	UnitID             int      `json:"unit_id"`
	EstimatedUnitPrice *float64 `json:"estimated_unit_price,omitempty"`
	Specifications     *string  `json:"specifications,omitempty"`
	SortOrder          int16    `json:"sort_order"`
}

type RFQAttachment struct {
	ID            int64     `json:"id"`
	RFQID         int64     `json:"rfq_id"`
	FileName      string    `json:"file_name"`
	FileURL       string    `json:"file_url"`
	FileSizeBytes *int64    `json:"file_size_bytes,omitempty"`
	UploadedBy    int64     `json:"uploaded_by"`
	UploadedAt    time.Time `json:"uploaded_at"`
}

type RFQVendorInvitation struct {
	RFQID      int64      `json:"rfq_id"`
	VendorID   int64      `json:"vendor_id"`
	InvitedAt  time.Time  `json:"invited_at"`
	NotifiedAt *time.Time `json:"notified_at,omitempty"`
}

type RFQSearchResult struct {
	Page       int   `json:"page"`
	Limit      int   `json:"limit"`
	Total      int   `json:"total"`
	TotalPages int   `json:"total_pages"`
	RFQs       []RFQ `json:"rfqs"`
}
